lib/FrontendTool/FrontendTool.cpp

```diff
@@ -871,6 +937,217 @@ static bool performCompile(CompilerInstance &Instance,
   return false;
 }
 
+/// If we are asked to link all, link all.
+static void linkAllIfNeeded(const CompilerInvocation &Invocation,
+                            SILModule *SM) {
+  if (Invocation.getSILOptions().LinkMode == SILOptions::LinkAll)
+    performSILLinking(SM, true);
+}
+
+/// Perform "stable" optimizations that are invariant across compiler versions.
+static bool performMandatorySILPasses(CompilerInvocation &Invocation,
+                                      SILModule *SM,
+                                      FrontendObserver *observer) {
+  if (Invocation.getFrontendOptions().RequestedAction ==
+      FrontendOptions::ActionType::MergeModules) {
+    // Don't run diagnostic passes at all.
+  } else if (!Invocation.getDiagnosticOptions().SkipDiagnosticPasses) {
+    if (runSILDiagnosticPasses(*SM))
+      return true;
+
+    if (observer) {
+      observer->performedSILDiagnostics(*SM);
+    }
+  } else {
+    // Even if we are not supposed to run the diagnostic passes, we still need
+    // to run the ownership evaluator.
+    if (runSILOwnershipEliminatorPass(*SM))
+      return true;
+  }
+
+  linkAllIfNeeded(Invocation, SM);
+
+  if (Invocation.getSILOptions().MergePartialModules)
+    SM->linkAllFromCurrentModule();
+  return false;
+}
+
+static SerializationOptions
+computeSerializationOptions(const CompilerInvocation &Invocation,
+                            bool moduleIsPublic) {
+  const FrontendOptions &opts = Invocation.getFrontendOptions();
+
+  SerializationOptions serializationOpts;
+  serializationOpts.OutputPath = opts.ModuleOutputPath.c_str();
+  serializationOpts.DocOutputPath = opts.ModuleDocOutputPath.c_str();
+  serializationOpts.GroupInfoPath = opts.GroupInfoPath.c_str();
+  if (opts.SerializeBridgingHeader)
+    serializationOpts.ImportedHeader = opts.ImplicitObjCHeaderPath;
+  serializationOpts.ModuleLinkName = opts.ModuleLinkName;
+  serializationOpts.ExtraClangOptions =
+      Invocation.getClangImporterOptions().ExtraArgs;
+  serializationOpts.EnableNestedTypeLookupTable =
+      opts.EnableSerializationNestedTypeLookupTable;
+  if (!Invocation.getIRGenOptions().ForceLoadSymbolName.empty())
+    serializationOpts.AutolinkForceLoad = true;
+
+  // Options contain information about the developer's computer,
+  // so only serialize them if the module isn't going to be shipped to
+  // the public.
+  serializationOpts.SerializeOptionsForDebugging =
+      !moduleIsPublic || opts.AlwaysSerializeDebuggingOptions;
+
+  return serializationOpts;
+}
+
+/// Perform SIL optimization passes if optimizations haven't been disabled.
+/// These may change across compiler versions.
+static void performSILOptimizations(CompilerInvocation &Invocation,
+                                    SILModule *SM) {
+  SharedTimer timer("SIL optimization");
+  if (Invocation.getFrontendOptions().RequestedAction ==
+          FrontendOptions::ActionType::MergeModules ||
```

Comment:
This isn't correct; MergeModules should run *no* passes.

Reply:
The old code *did* run that pass for MergeModules:
    // Perform SIL optimization passes if optimizations haven't been disabled.
    // These may change across compiler versions.
    {
      SharedTimer timer("SIL optimization");
      if (Action != FrontendOptions::ActionType::MergeModules &&
          Invocation.getSILOptions().shouldOptimize()) {

        runSILOptPreparePasses(*SM);

        StringRef CustomPipelinePath =
          Invocation.getSILOptions().ExternalPassPipelineFilename;
        if (!CustomPipelinePath.empty()) {
          runSILOptimizationPassesWithFileSpecification(*SM, CustomPipelinePath);
        } else {
          runSILOptimizationPasses(*SM);
        }
      } else {
        runSILPassesForOnone(*SM);
      }
    }
If I read the old code aright, then I would be making a functional change, which could conceivably delay batch mode testing. Am I missing something?

Reply:
Yuck. I believe that's incorrect behavior, but you're right, please leave it as is for now.

Reply:
Thanks!